### PR TITLE
feat: migrate hash func to func-v2

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

migrate hash func to func-v2

Closes #issue

### CODE REVIEW

File: src/query/functions-v2/src/scalars/hash.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+                output.commit_row();
+                Ok(())
+            },
+        ),
+    );
+
+    for ty in MAX_NUMERICS_TYPES {
+        with_number_mapped_type!(|NUM_TYPE| match ty {
+            NumberDataType::NUM_TYPE => {
+                registry.register_passthrough_nullable_2_arg::<StringType, NumberType<NUM_TYPE>, StringType, _, _>(
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/hash.rs:105]: `<StringType, Int64Type, StringType>` is enough here. 

The second arg is mostly constant.

File: src/query/functions-v2/src/scalars/hash.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        match self {
+            Scalar::Number(NumberScalar::Int64(v)) => DFHash::hash(v, state),
+            Scalar::Number(NumberScalar::UInt64(v)) => DFHash::hash(v, state),
+            Scalar::Number(NumberScalar::Float64(v)) => DFHash::hash(v, state),
+            Scalar::String(vals) => {
+                for v in vals {
+                    DFHash::hash(v, state);
+                }
+            }
+            _ => {}
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/hash.rs:306]: what if the other is uint8 ... ?

`with_number_mapped_type` could be used there.

File: src/query/functions-v2/src/scalars/hash.rs (hunk 1 above)
Comment (TCeason) [reply src/query/functions-v2/src/scalars/hash.rs:105]: I do a test, and we support this in function v1...

```sql
select sha2('x',224.0);
+----------------------------------------------------------+
| sha2('x', 224)                                           |
+----------------------------------------------------------+
| 54a2f7f92a5f975d8096af77a126edda7da60c5aa872ef1b871701ae |
+----------------------------------------------------------+

```

**Comment by TCeason** [comment]
In function v1, [DFHash](https://github.com/datafuselabs/databend/blob/cdf3bc22c097f0eed8bb79567a00628c63adfbee/src/query/functions/src/scalars/hashes/hash_base.rs#L189) has these:

impl<'a> DFHash for &'a VariantValue 
impl<'a> DFHash for ArrayValueRef<'a> 
impl<'a> DFHash for StructValueRef<'a> 

How to convert it into function-v2?

Need help cc @sundy-li  @andylokandy

**Review by sundy-li — APPROVED** [review]

**Pull request merged**